kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/format/KafkaEntryFormatter.java

```diff
@@ -61,15 +65,42 @@ public DecodeResult decode(List<Entry> entries, byte magic) {
 
         // reset header information
         final List<ByteBuf> orderedByteBuf = new ArrayList<>();
+
         for (Entry entry : entries) {
             try {
                 long startOffset = MessageIdUtils.peekBaseOffsetFromEntry(entry);
                 final ByteBuf byteBuf = entry.getDataBuffer();
                 final MessageMetadata metadata = Commands.parseMessageMetadata(byteBuf);
                 if (isKafkaEntryFormat(metadata)) {
+                    byte batchMagic = byteBuf.getByte(byteBuf.readerIndex() + MAGIC_OFFSET);
                     byteBuf.setLong(byteBuf.readerIndex() + OFFSET_OFFSET, startOffset);
-                    byteBuf.setByte(byteBuf.readerIndex() + MAGIC_OFFSET, magic);
-                    orderedByteBuf.add(byteBuf.slice(byteBuf.readerIndex(), byteBuf.readableBytes()));
+
+                    // batch magic greater than the magic corresponding to the version requested by the client
+                    // need down converted
+                    if (batchMagic > magic) {
+                        ConvertedRecords<MemoryRecords> convertedRecords = null;
+                        MemoryRecords memoryRecords = MemoryRecords.readableRecords(ByteBufUtils.getNioBuffer(byteBuf));
+                        //down converted, batch magic will be set to client magic
+                        convertedRecords = memoryRecords.downConvert(magic, startOffset, time);
+                        log.trace("[{}:{}] downConvert record, start offset {}, entry magic: {}, client magic: {}"
+                                , entry.getLedgerId(), entry.getEntryId(), startOffset, batchMagic, magic);
+                        if (convertedRecords != null) {
+                            final ByteBuf kafkaBuffer = Unpooled.wrappedBuffer(convertedRecords.records().buffer());
+                            orderedByteBuf.add(kafkaBuffer);
+                            if (!optionalByteBufs.isPresent()) {
+                                optionalByteBufs = Optional.of(new ArrayList<>());
+                            }
+                            optionalByteBufs.ifPresent(byteBufs -> byteBufs.add(byteBuf));
+                            optionalByteBufs.ifPresent(byteBufs -> byteBufs.add(kafkaBuffer));
+                            log.trace("[{}:{}] down convertedRecords not null {}, {}, {}"
+                                    , entry.getLedgerId(), entry.getEntryId(), startOffset, batchMagic, magic);
+                        } else {
+                            throw new IOException("DownConvert failed, convertedRecords is null");
+                        }
```

Comment:
AFAIK, `downConvert` never returns null. Even if it could return null, we should skip this entry because it means this entry is corrupted. If an exception was thrown for a corrupted entry, the exception would eventually be caught outside. However, the `entries` wouldn't be released in this case. See how we process the `KoPMessageMetadataNotFoundException`:

```java
            } catch (KoPMessageMetadataNotFoundException | IOException e) { // skip failed decode entry
                log.error("[{}:{}] Failed to decode entry. ", entry.getLedgerId(), entry.getEntryId(), e);
                entry.release();
            }
```

Reply:
@BewareMyPower Great find. Almost all exceptions in Kafka inherit from KafkaException and will be captured and processed in KafkaApis. We can catch KafkaException here. I updated the code, PTAL